### PR TITLE
Add a conda recipe for `nvjitlink`

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR adds a conda recipe for `nvjitlink`.

### CODE REVIEW

**Comment by brandon-b-miller** [comment]
Just an update here, this is WIP as the recipe currently does not build. I've merged the cmake changes from https://github.com/rapidsai/nvjitlink/pull/1 to test against and as of right now I end up with a conda build error where `nvJitLink_static` is not found:

<details><summary>conda build output</summary>
  error: subprocess-exited-with-error
  
  × Building wheel for nvjitlink (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [47 lines of output]
      *** scikit-build-core 0.4.7 using CMake 3.26.4 (wheel)
      *** Configuring CMake...
      2023-07-18 06:37:34,763 - scikit_build_core - WARNING - libdir/ldlibrary: /raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/libpython3.9.a is not a real file!
      2023-07-18 06:37:34,763 - scikit_build_core - WARNING - Can't find a Python library, got libdir=/raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib, ldlibrary=libpython3.9.a, multiarch=x86_64-linux-gnu, masd=None
      loading initial cache file build/cp39-cp39-manylinux_2_31_x86_64/CMakeInit.txt
      -- The CXX compiler identification is GNU 12.3.0
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: /raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_build_env/bin/x86_64-conda-linux-gnu-c++ - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Found Python: /raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/python3.9 (found version "3.9.16") found components: Interpreter Development Development.Module Development.Embed
      -- Unable to find cublas_v2.h in either "/raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_build_env/targets/x86_64-linux/include" or "/raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_build_env/math_libs/include"
      -- Found CUDAToolkit: /raid/brmiller/anaconda/conda-bld/nvjitlink_1689687111745/_build_env/targets/x86_64-linux/include (found suitable version "12.0.76", minimum required is "12.0")
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
      -- Looking for pthread_create in pthreads
      -- Looking for pthread_create in pthreads - not found
      -- Looking for pthread_create in pthread
      -- Looking for pthread_create in pthread - found
      -- Found Threads: TRUE
      -- Configuring done (2.3s)
      CMake Error at CMakeLists.txt:16 (target_link_libraries):
        Target "_nvjitlinklib" links to:
          CUDA::nvJitLink_static
        but the target was not found.  Possible reasons include:
          * There is a typo in the target name.
          * A find_package call is missing for an IMPORTED target.
          * An ALIAS target is missing.

</details>

At this time I'm still debugging this.

File: conda/recipes/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  path: ../../
+
+build:
+  noarch: python
+  script: pip install .
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda-nvcc') }}
```
Comment (bdice) [inline conda/recipes/meta.yaml:16]: Add a `conda_build_config.yaml` with these contents:
```yaml
c_compiler_version:
  - 11

cxx_compiler_version:
  - 11

cuda_compiler:
  - cuda-nvcc
```

And apply this suggestion:
```suggestion
    - {{ compiler('cuda') }}
```

File: conda/recipes/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    - cmake >=3.26.4
+  host:
+    - cuda-version=12.0
+    - python
+    - scikit-build-core
+    - libnvjitlink-dev
+    - libnvjitlink-static
+    - cuda-nvcc-impl
+  run:
+    - python
```
Comment (bdice) [inline conda/recipes/meta.yaml:26]:
```suggestion
    - python
    - {{ pin_compatible('cuda-version', max_pin='x', min_pin='x') }}
```

File: conda/recipes/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  imports:
+    - nvjitlink
+  source_files:
+    - nvjitlink/tests/test_nvjitlink_api.py
+    - nvjitlink/tests/test_nvjitlink.py
+    - nvjitlink/tests/test_device_functions.cubin
+    - nvjitlink/tests/test_device_functions.fatbin
+    - nvjitlink/tests/undefined_extern.cubin
+    
+
```
Comment (bdice) [inline conda/recipes/meta.yaml:39]:
```suggestion
```

**Review by bdice — COMMENTED** [review]
A couple small comments. Not sure about the CMake issue failing to find nvjitlink yet. Does this repo have CI scripts so that we can reproduce the failure?

File: conda/recipes/meta.yaml [hunk 4]
```diff
@@ -8,7 +8,9 @@ source:
 
 build:
   noarch: python
-  script: pip install . -vv
+  script:
+    - export CMAKE_PREFIX_PATH=${PREFIX}/targets/x86_64-linux/lib:$CMAKE_PREFIX_PATH
+    - pip install . -vv
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:13]: Just noting that this change was needed for CMake to find `libnvjitlink-static`

File: conda/recipes/meta.yaml (hunk 1 above)
Comment (vyasr) [reply conda/recipes/meta.yaml]: Out of curiosity, is there any reason to prefer this other than "This is what we've done everywhere else"? AFAIK the two should be equivalent (not including the version bits of course).

Comment (jakirkham) [reply conda/recipes/meta.yaml]: The `compiler('cuda')` syntax replaces this with the value from `cuda_compiler` (`cuda-nvcc` in Bradley's code), which in turn gets suffixed with the `target_platform` (becoming `cuda-nvcc_linux-64` for example), which is then the package name Conda uses to install

File: CMakeLists.txt [hunk 5]
```diff
@@ -9,10 +9,12 @@ project(
 find_package(Python COMPONENTS Interpreter Development REQUIRED)
 
 Python_add_library(_nvjitlinklib MODULE nvjitlink/_nvjitlinklib.cpp WITH_SOABI)
-
+set(CMAKE_FIND_DEBUG_MODE TRUE)
```
Comment (robertmaynard) [inline CMakeLists.txt:12]:
```suggestion
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -9,10 +9,12 @@ project(
...
 find_package(
   CUDAToolkit 12.0 REQUIRED
 )
+set(CMAKE_FIND_DEBUG_MODE FALSE)
```
Comment (robertmaynard) [inline CMakeLists.txt:16]:
```suggestion
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: conda/recipes/meta.yaml (hunk 4 above)
Comment (vyasr) [reply conda/recipes/meta.yaml:13]: Added a note

File: conda/recipes/meta.yaml (hunk 1 above)
Comment (vyasr) [reply conda/recipes/meta.yaml]: Right, I understand how it works but my question is whether there's any benefit to specifying `compiler('cuda')` and having the `cuda_compiler` in the CBC vs just using `compiler('cuda-nvcc')`. I know that's what we've done in our other recipes, and I'm fine with following that pattern for consistency, but I wanted to confirm that this was a stylistic suggestion and not functionally different.

File: conda/recipes/conda_build_config.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+c_compiler_version:
+  - 11 
+
+cxx_compiler_version: 
+  - 11 
+ 
+cuda_compiler: 
+  - cuda-nvcc
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:8]:
```suggestion
cuda_compiler: 
  - cuda-nvcc

cuda_version: 
  - 12.0
```

File: conda/recipes/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}
+    - cmake >=3.26.4
+    - sysroot_{{ target_platform }} 2.17
+    - cuda-version =12.0
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:26]:
```suggestion
    - cuda-version
```

File: conda/recipes/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  host:
+    - libnvjitlink-dev
+    - libnvjitlink-static
+    - cuda-version =12.0
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:30]:
```suggestion
    - cuda-version
```

File: conda/recipes/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    - python
+    - scikit-build-core
+  run:
+    - python
+    - numba >=0.57
+    - {{ pin_compatible('cuda-version', max_pin='x', min_pin='x') }}
+test:
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:37]: Would suggest separating these sections for readability

```suggestion
    - {{ pin_compatible('cuda-version', max_pin='x', min_pin='x') }}

test:
```

File: conda/recipes/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  imports:
+    - nvjitlink
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:39]: Did we want a smoke test like this with the extension module? Asking since it sounded like some errors got swallowed when `import`ing `nvjitlink`'s top-level package

```suggestion
    - nvjitlink._nvjitlink
    - nvjitlink
```

File: conda/recipes/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  source_files:
+    - nvjitlink/tests/test_nvjitlink_api.py
+    - nvjitlink/tests/test_nvjitlink.py
+    - nvjitlink/tests/test_device_functions.cubin
+    - nvjitlink/tests/test_device_functions.fatbin
+    - nvjitlink/tests/undefined_extern.cubin
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:45]: Is it worth adding some kind of comment about why these are copied over? In particular the `cubin` & `fatbin` files may be worth explaining

File: conda/recipes/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:46]: Looks like this is extraneous whitespace

```suggestion
```

File: conda/recipes/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    - nvjitlink/tests/test_device_functions.fatbin
+    - nvjitlink/tests/undefined_extern.cubin
+    
+  requires:
+    - pytest
+  commands: 
+    - pytest -v
+
+about:
+  home: https://github.com/rapidsai/nvjitlink
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:53]: Would flesh this out a bit. In particular packaging the license file. Pointing to related resources. And so on...

```suggestion
about:
  home: https://rapids.ai/
  dev_url: https://github.com/rapidsai/nvjitlink
  doc_url: https://github.com/rapidsai/nvjitlink/blob/main/README.md
  license: Apache-2.0
  license_family: Apache
  license_file: LICENSE
  license_url: https://github.com/rapidsai/nvjitlink/blob/main/LICENSE
  summary: nvJitLink Python binding
```

File: conda/recipes/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,53 @@
+package:
+  name: nvjitlink
+  version: 0.1.0
+
+source:
+  #url: https://github.com/rapidsai/nvjitlink/archive/refs/tags/0.1.0.tar.gz
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:6]: Sounds like we can drop this?

```suggestion
```

File: conda/recipes/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,53 @@
+package:
+  name: nvjitlink
+  version: 0.1.0
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:3]: Maybe we can template this somehow to ensure the version stays up-to-date with other sources of truth. This is one approach, but maybe there are others worth considering

```suggestion
{% set data = load_file_data("pyproject.toml") %}

package:
  name: nvjitlink
  version: {{ data.get("project", {}).get("version", "0.1.0.dev1") }}
```

File: conda/recipes/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  path: ../../
+
+build:
+  noarch: python
+  script:
+    # This is a workaround for library search path issues with the current
+    # cuda-nvcc package. They should be fixed by
+    # https://github.com/conda-forge/cuda-nvcc-feedstock/pull/24, at which
+    # point we can remove this.
+    - export CMAKE_PREFIX_PATH=${PREFIX}/targets/x86_64-linux/lib:$CMAKE_PREFIX_PATH
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:16]: xref: https://github.com/conda-forge/cuda-nvcc-feedstock/pull/24

File: conda/recipes/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    - {{ compiler('cuda') }}
+    - cmake >=3.26.4
+    - sysroot_{{ target_platform }} 2.17
+    - cuda-version =12.0
+  host:
+    - libnvjitlink-dev
+    - libnvjitlink-static
+    - cuda-version =12.0
+    - python
+    - scikit-build-core
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:32]: Think we want to list `pip` here explicitly since we are using it

```suggestion
    - python
    - pip
    - scikit-build-core
```

File: conda/recipes/meta.yaml [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+build:
+  noarch: python
+  script:
+    # This is a workaround for library search path issues with the current
+    # cuda-nvcc package. They should be fixed by
+    # https://github.com/conda-forge/cuda-nvcc-feedstock/pull/24, at which
+    # point we can remove this.
+    - export CMAKE_PREFIX_PATH=${PREFIX}/targets/x86_64-linux/lib:$CMAKE_PREFIX_PATH
+    - pip install . -vv
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:17]: Would recommend calling `pip` this way to ensure we are getting the right one from the right environment

```suggestion
    - {{ PYTHON }} -m pip install . -vv
```

File: CMakeLists.txt [hunk 20]
```diff
@@ -9,10 +9,10 @@ project(
 find_package(Python COMPONENTS Interpreter Development REQUIRED)
 
 Python_add_library(_nvjitlinklib MODULE nvjitlink/_nvjitlinklib.cpp WITH_SOABI)
-
 find_package(
   CUDAToolkit 12.0 REQUIRED
 )
+
```
Comment (jakirkham) [inline CMakeLists.txt:15]: Guessing these are left over from previous changes here. Maybe we can drop them unless we prefer them for some reason

**Review by jakirkham — COMMENTED** [review]
Thanks all! 🙏

Had some feedback below on things we can tighten up here (since it sounds like we are largely happy with how this is working)

File: conda/recipes/meta.yaml (hunk 15 above)
Comment (brandon-b-miller) [reply conda/recipes/meta.yaml:6]: I had thought that what would happen is that once satisfied here, we'd cut a release of `nvjitlink`, move this recipe to conda forge and uncomment this line pointing it to the release we want. Let me know if this sounds right.

File: conda/recipes/meta.yaml (hunk 11 above)
Comment (brandon-b-miller) [reply conda/recipes/meta.yaml:43]: Just an update on this, still trying to figure out if this is actually a problem or something scrambled just for me locally.

File: conda/recipes/meta.yaml (hunk 12 above)
Comment (brandon-b-miller) [reply conda/recipes/meta.yaml]: Added a note here

File: conda/recipes/meta.yaml (hunk 15 above)
Comment (jakirkham) [reply conda/recipes/meta.yaml:8]: Yeah that's ok

Still would drop this commented line from here

**Comment by jakirkham** [comment]
ok to test

File: conda/recipes/meta.yaml [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    - nvjitlink/tests/test_nvjitlink_api.py
+    - nvjitlink/tests/test_nvjitlink.py
+    # Extra files for testing edge cases of libnvjitlink
+    # see test_nvjitlink_api.py for details
+    - nvjitlink/tests/test_device_functions.cubin
+    - nvjitlink/tests/test_device_functions.fatbin
+    - nvjitlink/tests/undefined_extern.cubin
+  requires:
+    - pytest
+  commands:
```
Comment (jakirkham) [inline conda/recipes/meta.yaml:53]: Looks like the linter [caught some extra whitespace]( https://github.com/rapidsai/pynvjitlink/actions/runs/7493476170/job/20399259698?pr=2#step:6:95 )

```suggestion
  commands:
```

File: conda/recipes/conda_build_config.yaml [hunk 22]
```diff
@@ -0,0 +1,11 @@
+c_compiler_version:
+  - 11
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:2]:
```suggestion
  - 11
```

File: conda/recipes/conda_build_config.yaml [hunk 23]
```diff
@@ -0,0 +1,11 @@
+c_compiler_version:
+  - 11 
+
+cxx_compiler_version:
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:4]:
```suggestion
cxx_compiler_version:
```

File: conda/recipes/conda_build_config.yaml [hunk 24]
```diff
@@ -0,0 +1,11 @@
+c_compiler_version:
+  - 11 
+
+cxx_compiler_version: 
+  - 11
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:5]:
```suggestion
  - 11
```

File: conda/recipes/conda_build_config.yaml [hunk 25]
```diff
@@ -0,0 +1,11 @@
+c_compiler_version:
+  - 11 
+
+cxx_compiler_version: 
+  - 11 
+
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:6]:
```suggestion

```

File: conda/recipes/conda_build_config.yaml [hunk 26]
```diff
@@ -0,0 +1,11 @@
+c_compiler_version:
+  - 11 
+
+cxx_compiler_version: 
+  - 11 
+ 
+cuda_compiler: 
+  - cuda-nvcc
+
+cuda_version:
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:10]:
```suggestion
cuda_version:
```

File: conda/recipes/conda_build_config.yaml [hunk 27]
```diff
@@ -0,0 +1,10 @@
+c_compiler_version:
+  - 11
+
+cxx_compiler_version:
+  - 11
+cuda_compiler:
```
Comment (jakirkham) [inline conda/recipes/conda_build_config.yaml:6]:
```suggestion

cuda_compiler:
```

**Comment by bdice** [comment]
Closing in favor of #33.

**Comment by jakirkham** [comment]
Sorry for the noise
